src/common/metrics/ngx_wa_histogram.c

```diff
@@ -160,13 +180,24 @@ histogram_log(ngx_wa_metrics_t *metrics, ngx_wa_metric_t *m, uint32_t mid)
 
 
 ngx_int_t
-ngx_wa_metrics_histogram_add_locked(ngx_wa_metrics_t *metrics,
-    ngx_wa_metric_t *m)
+ngx_wa_metrics_histogram_add_locked(ngx_wa_metrics_t *metrics, uint32_t *bins,
+    uint16_t cn_bins, ngx_wa_metric_t *m)
 {
-    size_t                        i;
-    static uint16_t               n_bins = NGX_WA_METRICS_BINS_INIT;
+    size_t                        i, j = 0;
+    uint16_t                      n_bins = NGX_WA_METRICS_BINS_INIT;
+    ngx_wa_histogram_type_e       h_type = NGX_WA_HISTOGRAM_LOG2;
     ngx_wa_metrics_histogram_t  **h;
 
+    if (bins) {
+        if (cn_bins > NGX_WA_METRICS_BINS_MAX) {
```

Comment:
One last thing before merge: a logarithmic histogram can contain up to 18 bins as per the existing design, but this specific condition and the extra bin on line 197 mean that a custom binning histogram can contain up to 19 bins (18 defined + the MAX_UINT32 bin). Is this correct?

Reply:
I made this change in the merge PR. The change makes the documentation easier to read and understand: "all histograms can only have 18 bins"

Reply:
Thanks for addressing it in the merge PR.